### PR TITLE
GEODE-8020: buffer corruption in SSL communications

### PR DESCRIPTION
**Pull request by bschuchardt**
revert change in GEODE-6661 that made NioSslEngine use a direct buffer.
This class is not designed to share its buffer with a pool in
BufferPool.  Connection is also modified to use a heap buffer for
reading encrypted SSL packets for consistency.  New tests ensure that
these buffers are the correct type when using SSL or plain sockets.

Thank you for submitting a contribution to Apache Geode.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced in the commit message?

- [x] Has your PR been rebased against the latest commit within the target branch (typically `develop`)?

- [x] Is your initial contribution a single, squashed commit?

- [x] Does `gradlew build` run cleanly?

- [x] Have you written or updated unit tests to verify your changes?

- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?

### Note:
Please ensure that once the PR is submitted, check Concourse for build issues and
submit an update to your PR as soon as possible. If you need help, please send an
email to dev@geode.apache.org.

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/internal/tcp/Connection.java [hunk 1]
```diff
@@ -1763,9 +1763,13 @@ private static boolean isIgnorableIOException(Exception e) {
     }
 
     msg = msg.toLowerCase();
-    return msg.contains("forcibly closed")
-        || msg.contains("reset by peer")
-        || msg.contains("connection reset");
+
+    if (e instanceof SSLException && msg.contains("status = closed")) {
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/internal/tcp/Connection.java:1767]: The commit message references reverting the direct buffer changes. Is this change intentional because it doesn't look like part of the reverted commit?

File: geode-core/src/main/java/org/apache/geode/internal/tcp/Connection.java [hunk 2]
```diff
@@ -1662,8 +1662,8 @@ private void readMessages() {
         } catch (IOException e) {
           // "Socket closed" check needed for Solaris jdk 1.4.2_08
           if (!isSocketClosed() && !"Socket closed".equalsIgnoreCase(e.getMessage())) {
-            if (logger.isDebugEnabled() && !isIgnorableIOException(e)) {
-              logger.debug("{} io exception for {}", p2pReaderName(), this, e);
+            if (logger.isInfoEnabled() && !isIgnorableIOException(e)) {
```
[the remark this answers is not in the document]
Comment (bschuchardt) [reply geode-core/src/main/java/org/apache/geode/internal/tcp/Connection.java:1665]: yes!  I only saw the SSLException that explained the message-loss when debug-level logging was enabled.  With this log level change I had to augment isIgnorableException to filter out some exceptions being caught here when the socket was closed.

File: geode-core/src/main/java/org/apache/geode/internal/tcp/Connection.java (hunk 1 above)
Comment (bschuchardt) [reply geode-core/src/main/java/org/apache/geode/internal/tcp/Connection.java:1767]: I'm confused by this comment because I very clearly included changing back to non-direct buffers in Connection and NioSslEngine in this commit.
